### ISSUE TITLE
[🐛 Bug]: Dotnet Error in 4.11.0 when using ChromeDriverService.CreateDefaultService()

### ISSUE THREAD

**Issue by ZianAtFirstWatch** [SeleniumHQ/selenium#12564]
### What happened?

When using ChromeDriverService.CreateDefaultService() to instantiate a ChromeDriver, the error "Unable to obtain chrome using Selenium Manager" occurs. This started happening in 4.11.0, but was fine in 4.10.0.

### How can we reproduce the issue?

```shell
# This PowerShell code should be saved as a file with a ".ps1" extension. It uses the C# Nuget Packages' DLLs.
# They can be obtained by downloading the Nuget packages as ZIP files (https://www.nuget.org/packages/Selenium.Support/ and https://www.nuget.org/packages/Selenium.WebDriver) and copying the DLLs from the lib\net45 folder inside each ZIP file.

Function New-ChromeDriver {
    [CmdletBinding(
        SupportsShouldProcess,
        ConfirmImpact = 'Low'
    )]
    param(
        [Parameter(Mandatory)][string]$ScriptPath
    )
    $private:result = $null
    Initialize-Selenium -ScriptPath $ScriptPath
    $private:chromeDriverService = [OpenQA.Selenium.Chrome.ChromeDriverService]::CreateDefaultService()
    $private:chromeDriverService.DriverServicePath = $ScriptPath

    $private:chromeOptions = New-Object OpenQA.Selenium.Chrome.ChromeOptions
    $private:chromeOptions.AddExcludedArguments("excludeSwitches", "enable-logging");

    if ($PSCmdlet.ShouldProcess("Create a new instance of ChromeDriver.")) {
        $private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeDriverService, $private:chromeOptions)
    }

    $private:result
}

Function Initialize-Selenium([string]$ScriptPath) {
    #Required for Chrome
    if (($env:Path -split ';') -notcontains $ScriptPath) {
        $env:Path += ";$ScriptPath"
    }

    Import-Module "$($ScriptPath)\WebDriver.dll"
    Import-Module "$($ScriptPath)\WebDriver.Support.dll"
}

$FolderWithDllsAndChromeDriverExecutable = Read-Host -Prompt "Which folder has the WebDriver DLLs and ChromeDriver executable file?"
New-ChromeDriver -ScriptPath $FolderWithDllsAndChromeDriverExecutable
```

### Relevant log output

```shell
https://www.selenium.dev/documentation/webdriver/troubleshooting/logging/ says ".NET does not currently have a Logging implementation".
```

### Operating System

Windows 10

### Selenium version

4.11.0

### What are the browser(s) and version(s) where you see this issue?

Version 116.0.5845.97 (Official Build) (64-bit)

### What are the browser driver(s) and version(s) where you see this issue?

ChromeDriver 116.0.5845.96 (1a391816688002153ef791ffe60d9e899a71a037-refs/branch-heads/5845@{#1382})

### Are you using Selenium Grid?

No

**Comment by titusfortner** [comment]
I suspect the value of $ScriptPath does not have a trailing slash to indicate that it is a directory and not a file.
That bug has been fixed for the next release.
Further, you shouldn't need to specify a service any longer, since selenium will manage the drivers for you.

**Comment by ZianAtFirstWatch** [comment]
Thanks for the good news about the next release.

Is there a GitHub issue # or some other URL I can visit to see the change?

>What about when a trailing slash is included?

PS C:\Users\zchoy\Downloads\test copy> .\test.ps1
Which folder has the WebDriver DLLs and ChromeDriver executable file?: C:\Users\zchoy\Downloads\test copy\
MethodInvocationException: C:\Users\zchoy\Downloads\test copy\test.ps1:11
Line |
  11 |      $private:chromeDriverService = [OpenQA.Selenium.Chrome.ChromeDriv …
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling "CreateDefaultService" with "0" argument(s): "Unable to obtain chrome using Selenium Manager;
     | For documentation on this error, please visit:
     | https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location"

InvalidOperation: C:\Users\zchoy\Downloads\test copy\test.ps1:12
Line |
  12 |      $private:chromeDriverService.DriverServicePath = $ScriptPath
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The property 'DriverServicePath' cannot be found on this object. Verify that the property exists and can be set.

New-Object: C:\Users\zchoy\Downloads\test copy\test.ps1:18
Line |
  18 |  … te:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver  …
     |                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Exception calling ".ctor" with "2" argument(s): "Object reference not set to an instance of an object."

**Comment by titusfortner** [comment]
.NET is frustrating because we don't have good logging for what is going on.
The Service class constructor takes a String argument representing the path of the directory that the driver is in. 
In Selenium 4.10, if it didn't find the driver in that directory it would ignore the whole thing and look on PATH.

In 4.11, Selenium doesn't ignore it, and if the driver isn't there it fails. The problem is that people were putting the full path including the file name, which was failing. So I added code to split up directory and file name automatically if someone passed in a file and not a directory. Except, .NET assumes that if there is no trailing slash that it is a file and not a directory, which broke things.

So. 4.11 you don't need to specify anything; delete any drivers on the system and Selenium should download everything that needs to be downloaded based on the Options class values. If there is a problem and you need to specify the location, make sure you are passing in only the directory that the driver is located in and use a trailing slash.

In 4.12 the conditional will be improved so that it figures out if it is a file first rather than assuming the lack of a trailing slash makes it a file and not a directory.

**Comment by ChrisBloom-TH** [comment]
Seeing this same output error on my side **intermittently** with the following log after a simple startup. No chromedriver exists in the bin directory / on any system path, we're relying on Selenium Manager to download the driver at runtime. Hopefully this log is potentially a bit more helpful.

Sample code
                    ChromeOptions options = new ChromeOptions();
                   
                    options.AddArguments("-auth-server-whitelist=*.companyurl.com");
                    options.AddArguments("-auth-negotiate-delegatewhitelist=*.companyurl.com");
                    options.AddArguments("-auth-schemes=digest,ntlm,negotiate");

                  options.AddUserProfilePreference("download.default_directory", @"C:\Automation\VSTS\Other\Downloads");

                    newDriver = new ChromeDriver(ChromeDriverService.CreateDefaultService(), options, TimeSpan.FromSeconds(180));




DEBUG[13:22:12.252]: Exception Starting Chrome Driver OpenQA.Selenium.NoSuchDriverException: **Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location** ---> OpenQA.Selenium.WebDriverException: Error starting process: C:\Automation\VSTS\SeleniumDriver\bin\Release\selenium-manager/windows/selenium-manager.exe  --browser "chrome" --output json ---> OpenQA.Selenium.WebDriverException: **Selenium Manager process exited abnormally with 65 code**: C:\Automation\VSTS\SeleniumDriver\bin\Release\selenium-manager/windows/selenium-manager.exe  --browser "chrome" --output json
{
  "logs": [
    {
      "level": "ERROR",
      "timestamp": 1692292932,
      "message": "error sending request for url (https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json): error trying to connect: An existing connection was forcibly closed by the remote host. (os error 10054)"
    }
  ],
  "result": {
    "code": 65,
    "message": "error sending request for url (https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json): error trying to connect: An existing connection was forcibly closed by the remote host. (os error 10054)",
    "driver_path": "",
    "browser_path": ""
  }
}


   at OpenQA.Selenium.SeleniumManager.RunCommand(String fileName, String arguments)
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.SeleniumManager.RunCommand(String fileName, String arguments)
   at OpenQA.Selenium.SeleniumManager.DriverPath(DriverOptions options)
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   at OpenQA.Selenium.Chrome.ChromeDriverService.CreateDefaultService(ChromeOptions options)
   at OpenQA.Selenium.Chrome.ChromeDriverService.CreateDefaultService()
   at KeyWordDriver.Selenium.DriverHelper.CreateRemoteDriver(Boolean TempDriver, Boolean skipIfMobile) in C:\VSTS\_work\7\s\SeleniumDriver\Selenium\DriverHelper.cs:line 1440

**Comment by ZianAtFirstWatch** [comment]
> .NET is frustrating because we don't have good logging for what is going on. The Service class constructor takes a String argument representing the path of the directory that the driver is in. In Selenium 4.10, if it didn't find the driver in that directory it would ignore the whole thing and look on PATH.
> 
> In 4.11, Selenium doesn't ignore it, and if the driver isn't there it fails. The problem is that people were putting the full path including the file name, which was failing. So I added code to split up directory and file name automatically if someone passed in a file and not a directory. Except, .NET assumes that if there is no trailing slash that it is a file and not a directory, which broke things.
> 
> So. 4.11 you don't need to specify anything; delete any drivers on the system and Selenium should download everything that needs to be downloaded based on the Options class values. If there is a problem and you need to specify the location, make sure you are passing in only the directory that the driver is located in and use a trailing slash.
> 
> In 4.12 the conditional will be improved so that it figures out if it is a file first rather than assuming the lack of a trailing slash makes it a file and not a directory.

Unfortunately, the error happens even when $env:path contains the path of the directory with a trailing slash and without the name of the file. The path is the last entry in the following text.

`C:\Program Files\PowerShell\7;C:\Program Files\Eclipse Adoptium\jdk-11.0.18.10-hotspot\bin;C:\Program Files\Microsoft MPI\Bin\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\PROGRAMDATA\WEBEX\WEBEX\319\;C:\PROGRAMDATA\WEBEX\;C:\Program Files\SourceGear\Common\DiffMerge\;c:\Program Files (x86)\Microsoft SQL Server\90\Tools\binn\;C:\Program Files\dotnet\;C:\Program Files\Microsoft SQL Server\130\Tools\Binn\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\Azure Data Studio\bin;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn\;C:\Program Files\Liquid Technologies\Liquid Studio 2021\XmlDataBinder19\Redist19\cpp\win32\bin;C:\Program Files\Liquid Technologies\Liquid Studio 2021\XmlDataBinder19\Redist19\cpp\win64\bin;C:\Program Files\PuTTY\;C:\Program Files\Liquid Technologies\Liquid Studio\XmlDataBinder20\Redist20\cpp\win32\bin;C:\Program Files\Liquid Technologies\Liquid Studio\XmlDataBinder20\Redist20\cpp\win64\bin;C:\PROGRAM FILES (X86)\WEBEX\WEBEX\319\;C:\PROGRAM FILES (X86)\WEBEX\;C:\Program Files (x86)\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\Intel\Intel(R) Management Engine Components\DAL;C:\Program Files\TortoiseHg\;C:\Program Files (x86)\Microsoft SQL Server\160\DTS\Binn\;C:\Program Files\PowerShell\7\;C:\Users\zchoy\AppData\Local\Microsoft\WindowsApps;C:\Users\zchoy\AppData\Local\Programs\Fiddler;C:\Users\zchoy\AppData\Local\GitHubDesktop\bin;C:\Users\zchoy\AppData\Local\Microsoft\WindowsApps;C:\Users\zchoy\.dotnet\tools;C:\Program Files\Azure Data Studio\bin;C:\Users\zchoy\documents\computer issues\sysinternals;;C:\Users\zchoy\Downloads\test copy\`

**Comment by titusfortner** [comment]
@ChrisBloom-TH that's a different error from this one. Selenium Manager should be catching issues with not connecting and using cache if found. @bonigarcia what causes the code 65 there? Anything SM can do to address it?

@ZianAtFirstWatch I'm not referring to `$env:path`, I'm referring to `$ScriptPath`.

**Comment by bonigarcia** [comment]
Error code 65 is thrown by Selenium Manager when an exception is caught in the logic execution. It means "data error" and typically is caused by some inconsistent error in the user data. In this case, it seems to be a connectivity issue:

```
"error sending request for url (https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json): error trying to connect: An existing connection was forcibly closed by the remote host. (os error 10054)"
```

Maybe some proxy setup is required.

**Comment by titusfortner** [comment]
@bonigarcia can we capture/ignore this error the same way we do if the system is offline?

**Comment by titusfortner** [comment]
(if there is something already cached that works, we should be able to use that)

**Comment by bonigarcia** [comment]
> (if there is something already cached that works, we should be able to use that)

When the driver is already in the cache, it is used by Selenium Manager.

> @bonigarcia can we capture/ignore this error the same way we do if the system is offline?

In this issue, it seems there is a connectivity problem: `An existing connection was forcibly closed by the remote host`. I believe it is a legitimate error, and we should not ignore it.

**Comment by titusfortner** [comment]
Aren't we checking `known-good-versions-with-downloads.json` before we know if there's something in the cache?

**Comment by bonigarcia** [comment]
Eventually, yes. That endpoint should be read to refresh the metadata (i.e., the resolved driver version), ensuring that what's in the cache is still valid. The point is what you said "_if there is something already cached **that works**_". We need to ask the endpoints again to ensure that the cache works.

A fallback mechanism is to use the latest driver in the cache when there is some problem (such as connectivity or any other exception). I implemented this fallback in WebDriverManager but not in Selenium Manager because, precisely, it is a fallback, i.e., it is not the optimal solution.

For instance, the optimal solution to this issue would be to detect what is causing the intermittent connectivity issues. My guess is that a problem which is related to the network (e.g., because of a proxy or firewall) on the user side.

Being said that. If we want, we can implement that fallback (i.e., get whatever the last driver is on the cache). It could be the default way of working for the offline mode. Also, in the online (i.e., default) mode, Selenium Manager can retry in offline mode if some error happens. It is a fallback, not the optimal solution, but maybe it works for a number of cases.

**Comment by titusfortner** [comment]
> A fallback mechanism is to use the latest driver in the cache when there is some problem (such as connectivity or any other exception)

I don't think this is even necessarily a fallback, though. I may want to run my tests against localhost without being online. I shouldn't need to be online if I have something valid cached.

I think we should throw a warning if there's a problem in our normal operations and still provide a valid driver whenever we can.

Maybe it is a real problem and they will have to figure it out when the browser version gets updated next, or maybe it is a weird intermittent issue that hopefully no one ever has to look at because we accommodate for it.

**Comment by StevenTN615** [comment]
I'm not trying to hijack this thread, but this message is specifically for titusfortner.  You said in [this reply](https://github.com/SeleniumHQ/selenium/issues/12564#issuecomment-1681528698) to the OP

> Further, you shouldn't need to specify a service any longer, since selenium will manage the drivers for you.

I can get the OP's PowerShell code to work if I pass the full path (with the trailing slash) to the ChromeDriverService in it's constructor . However, if I comment out those lines for creating the ChromeDriverService, and only attempt to create the ChromeDriver passing in the ChromeOptions exclusively such as:

_$private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeOptions)_

I end up getting the following exception:

_New-Object : Exception calling ".ctor" with "1" argument(s): "Unable to obtain chrome using Selenium Manager_

What am I missing if I don't need to specify the service any more? I would like to have Selenium keep my chromedriver.exe updated automatically, instead of having to use a separate script to do it.

I can manually execute the Selenium-Manager CLI to create the proper chromedriver.exe in the .cache folder, but the WebDriver.dll doesn't seem to do it for me automatically. I'm using the 4.11 WebDriver.dll for .Net 4.8.

Thanks.

**Comment by titusfortner** [comment]
@diemol the primary issue with this is that the binary itself doesn't adjust to the fact that the user might be offline when trying to run the test with a valid driver already in the cache. I'm a little confused about how we are doing ttl vs checking the current driver version and browser version. If they match we don't need to call out again to update things.

@StevenTN615 I think this fixes your issue: https://github.com/SeleniumHQ/selenium/pull/12534

**Comment by StevenTN615** [comment]
@titusfortner Thanks for getting back to me. I did look at that link you sent, but still am having issues. I tried loading up the DriverFinder, but I then get an exception:

_Cannot find type [OpenQA.Selenium.Remote.Service.DriverFinder]: verify that the assembly containing this type is 
loaded._

I'm not sure what assembly I need to load that contains the DriverFinder. I'm just trying to update some PowerShell code to use Selenium 4.11 and wanted the browser driver to automatically get found and loaded instead of manually doing it. So confused.

**Comment by titusfortner** [comment]
@StevenTN615 sorry, I meant you need to wait for me to merge that PR and release Selenium 4.12. I'm planning to do that Monday if everything goes well.

**Comment by StevenTN615** [comment]
@titusfortner Thanks, I knew I was confused for a reason. Looking forward to trying Selenium 4.12.

**Comment by nuthanbanad** [comment]
@titusfortner I am using the same Selenium version 4.11.0 and chrome browser version 116.0.5845.140 and getting same error message as below when running my C# selenium automation scripts intermittently.

"Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location "

>>  I observed another behavior that this issue is also happening when the chrome browser is in updating state.. and works fine once browser update is finished.

Below is my code to instantiate chromedriver. Please let me know If I am doing anything wrong.
          service = ChromeDriverService.CreateDefaultService();
          service.HideCommandPromptWindow = true;
          ChromeOptions options = new ChromeOptions();
          options.AddArgument("no-sandbox");
          options.AddArguments("--headless");
          options.AddArgument("--window-size=1920,1080");
          options.AddUserProfilePreference("disable-popup-blocking", "true");
          chromeDriver = new ChromeDriver(service, options, TimeSpan.FromMinutes(3));

**Comment by titusfortner** [comment]
Need to wait for 4.12 to make sure things are good.

**Comment by nvborisenko** [comment]
@nuthanbanad please try the latest 4.12.4 and share full exception output if any. Thanks.

**Comment by StevenTN615** [comment]
I downloaded the C# package for 4.12.4 and extracted the webdriver.dll. When I run the OP's code for PowerShell and remove the instantiation of the ChromeDriverService and only try to create the ChromeDriver object passing in only the ChromeOptions, I still end up with the same error as with 4.11.

New-Object : Exception calling ".ctor" with "1" argument(s): "Unable to obtain chrome using Selenium 
Manager; For documentation on this error, please visit: 
https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location"
At C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\ZZZSeleniumTest2.ps1:31 char:27
+ ... te:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver  ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [New-Object], MethodInvocationException
    + FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCom 
   mand

I don't know what I'm missing here in order to have the Selenium manager find the correct  version of ChromeDriver.exe that I require. 

Everything works fine if I manually put the chromedriver.exe in my script path, and create the ChromeDriverService, but I thought the whole point on the newer Selenium is to not have to create the DriverService.

```powershell
Function New-ChromeDriver {
    [CmdletBinding(
        SupportsShouldProcess,
        ConfirmImpact = 'Low'
    )]
    param(
        [Parameter(Mandatory)][string]$ScriptPath
    )
    $private:result = $null
    Initialize-Selenium -ScriptPath $ScriptPath
    #$private:chromeDriverService = [OpenQA.Selenium.Chrome.ChromeDriverService]::CreateDefaultService($ScriptPath)


    $private:chromeOptions = New-Object OpenQA.Selenium.Chrome.ChromeOptions
    $private:chromeOptions.AddExcludedArguments("excludeSwitches", "enable-logging");

    if ($PSCmdlet.ShouldProcess("Create a new instance of ChromeDriver.")) {
        #$private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeDriverService, $private:chromeOptions)
        $private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeOptions)
    }

    $private:result


}

Function Initialize-Selenium([string]$ScriptPath) {
    Import-Module "$($ScriptPath)WebDriver.dll"
}

$FolderWithDllsAndChromeDriverExecutable = "C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\"
New-ChromeDriver -ScriptPath $FolderWithDllsAndChromeDriverExecutable
```

**Comment by nvborisenko** [comment]
Ok, you unpack the package yourself, without any package manager.

Then you should make sure the following folders structure:
```
Selenium.WebDriver.dll
selenium-manager/windows/selenium-manager.exe
```

So, you are missing `selenium-manager.exe`.

**Comment by StevenTN615** [comment]
Thanks for getting back to me, but my ignorance is still getting the best of me.

I have WebDriver.dll located in the following directory:
       C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\

Should selenium-manager.exe  be added in the following directory:
      C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\selenium-manager\windows\

Or is the directory not relative to where WebDriver.dll is located?

I've tried it by putting selenium-manager.exe in 
      C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\selenium-manager\windows\
but I still get the same exception as above.

**Comment by nvborisenko** [comment]
It should be
```
C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\Selenium.WebDriver.dll
C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\selenium-manager\windows\selenium-manager.exe
```

Just to have clean discussion and understanding where we are, please share now your code you use, and exception you get.

**Comment by StevenTN615** [comment]
I'm basically using the same code as the OP, but I hard coded the path to WebDriver.dll and I removed the instantiation of the Chrome Driver Service since I keep getting told it is no longer required.  I also renamed WebDriver.dll to Selenium.WebDriver.dll since that is what you showed in your example (although I don't think that should matter. 

I also put selenium-manager.exe in the folder structure that you specified.

Thanks for your patience.

Still get the same exception:

```powershell
New-Object : Exception calling ".ctor" with "1" argument(s): "Unable to obtain chrome using Selenium 
Manager; For documentation on this error, please visit: 
https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location"
At C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\ZZZSeleniumTest3.ps1:16 char:27
+ ... te:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver  ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [New-Object], MethodInvocationException
    + FullyQualifiedErrorId : ConstructorInvokedThrowException,Microsoft.PowerShell.Commands.NewObjectCom 
   mand

``` 

```powershell
Function New-ChromeDriver {
    [CmdletBinding(
        SupportsShouldProcess,
        ConfirmImpact = 'Low'
    )]
    param(
        [Parameter(Mandatory)][string]$ScriptPath
    )

    $private:result = $null
    Initialize-Selenium -ScriptPath $ScriptPath
    $private:chromeOptions = New-Object OpenQA.Selenium.Chrome.ChromeOptions
    $private:chromeOptions.AddExcludedArguments("excludeSwitches", "enable-logging");

    if ($PSCmdlet.ShouldProcess("Create a new instance of ChromeDriver.")) {
        $private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeOptions)
    }

    $private:result
}

Function Initialize-Selenium([string]$ScriptPath) {
    Import-Module "$($ScriptPath)Selenium.WebDriver.dll"
}

$FolderWithDllsAndChromeDriverExecutable = "C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\"
#$FolderWithDllsAndChromeDriverExecutable = Read-Host -Prompt "Which folder has the WebDriver DLLs and ChromeDriver executable file?"
New-ChromeDriver -ScriptPath $FolderWithDllsAndChromeDriverExecutable
```

**Comment by nvborisenko** [comment]
Weird thing here we don't see full exception stack trace including inner exceptions. Do you know how to poll it out in powershell?

**Comment by StevenTN615** [comment]
Not particularly, but I added the following try/catch block around when I create the ChromeDriver object.

```powershell
        try {
            $private:result = New-Object -TypeName OpenQA.Selenium.Chrome.ChromeDriver -ArgumentList @($private:chromeOptions)
        } catch {
            $Error[0].Exception.StackTrace
            Write-Host " "
            $Error[0].Exception.InnerException.StackTrace
            Write-Host " "
            $Error[0].StackTrace
        }
``` 

This is the output, if that helps.

```
   at System.Management.Automation.DotNetAdapter.AuxiliaryConstructorInvoke(MethodInformation methodInforma
tion, Object[] arguments, Object[] originalArguments)
   at System.Management.Automation.DotNetAdapter.ConstructorInvokeDotNet(Type type, ConstructorInfo[] const
ructors, Object[] arguments)
   at Microsoft.PowerShell.Commands.NewObjectCommand.CallConstructor(Type type, ConstructorInfo[] construct
ors, Object[] args)
 
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   at OpenQA.Selenium.Chromium.ChromiumDriver.GenerateDriverServiceCommandExecutor(DriverService service, D
riverOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chromium.ChromiumDriver..ctor(ChromiumDriverService service, ChromiumOptions options,
 TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeDriverService service, ChromeOptions options, TimeSpa
n commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeOptions options)

```

**Comment by nvborisenko** [comment]
Much better, let's improve to see inner exceptions:
```
$Error[0].Exception.ToString()
```

I mean execute `.ToString()` method which will show us what we want.

**Comment by StevenTN615** [comment]
Okay. Output here:

```
System.Management.Automation.MethodInvocationException: Exception calling ".ctor" with "1" argument(s): "Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location" ---> OpenQA.Selenium.NoSuchDriverException: Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: 
https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location ---> System.TypeInitializationException: The type initializer for 'OpenQA.Selenium.SeleniumManager' threw an exception. ---> OpenQA.Selenium.WebDriverException: Unable to locate or obtain Selenium Manager binary at C:\WINDOWS\system32\WindowsPowerShell\v1.0\selenium-manager/windows/selenium-manager.exe
   at OpenQA.Selenium.SeleniumManager..cctor()
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.SeleniumManager.DriverPath(DriverOptions options)
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   at OpenQA.Selenium.Chromium.ChromiumDriver.GenerateDriverServiceCommandExecutor(DriverService service, DriverOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chromium.ChromiumDriver..ctor(ChromiumDriverService service, ChromiumOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeDriverService service, ChromeOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeOptions options)
   --- End of inner exception stack trace ---
   at System.Management.Automation.DotNetAdapter.AuxiliaryConstructorInvoke(MethodInformation methodInformation, Object[] arguments, Object[] originalArguments)
   at System.Management.Automation.DotNetAdapter.ConstructorInvokeDotNet(Type type, ConstructorInfo[] constructors, Object[] arguments)
   at Microsoft.PowerShell.Commands.NewObjectCommand.CallConstructor(Type type, ConstructorInfo[] constructors, Object[] args)
 
``` 

I see that it says it can't find selenium-manager.exe @ _C:\WINDOWS\system32\WindowsPowerShell\v1.0\selenium-manager/windows/selenium-manager.exe_, so I added it exactly as the path states into the system32 sub directory. Still gives the same exception. Could it be that the path after selenium-manager all the \ become /?

**Comment by StevenTN615** [comment]
@nvborisenko  I think I was mixing the 4.11 WebDriver.dll with the 4.12 selenium-manager.exe, so I deleted everything, re-pulled the 4.12.4 C# package, unpacked the package, and moved the WebDriver.dll to my test directory ("C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\"), and the selenium-manager.exe to "C:\WINDOWS\system32\WindowsPowerShell\v1.0\selenium-manager/windows/" and re-ran my test script. Now I get a different exception, so that's a step in the right direction.

```
System.Management.Automation.MethodInvocationException: Exception calling ".ctor" with "1" argument(s): "Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: 
https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location" 
---> OpenQA.Selenium.NoSuchDriverException: Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: 
https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location 
---> System.IO.FileNotFoundException: Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified.
   at OpenQA.Selenium.SeleniumManager.RunCommand(String fileName, String arguments)
   at OpenQA.Selenium.SeleniumManager.DriverPath(DriverOptions options)
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   --- End of inner exception stack trace ---
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   at OpenQA.Selenium.Chromium.ChromiumDriver.GenerateDriverServiceCommandExecutor(DriverService service, DriverOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chromium.ChromiumDriver..ctor(ChromiumDriverService service, ChromiumOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeDriverService service, ChromeOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeOptions options)
   --- End of inner exception stack trace ---
   at System.Management.Automation.DotNetAdapter.AuxiliaryConstructorInvoke(MethodInformation methodInformation, Object[] arguments, Object[] originalArguments)
   at System.Management.Automation.DotNetAdapter.ConstructorInvokeDotNet(Type type, ConstructorInfo[] constructors, Object[] arguments)
   at Microsoft.PowerShell.Commands.NewObjectCommand.CallConstructor(Type type, ConstructorInfo[] constructors, Object[] args)


``` 

I even installed the newtonsoft.json module into PowerShell, and imported it into my script, but still get the same exception. 🤷‍♂️

**Comment by nvborisenko** [comment]
You need place `Newtonsoft.Json` dll beside `Selenium.WebDriver.dll`. And all other dlls as dependencies of Selenium.

You can explorer dependencies here: https://www.nuget.org/packages/Selenium.WebDriver#dependencies-body-tab

**Comment by StevenTN615** [comment]
Thank you, thank you, thank you. This now works as it should.

The only difference from what you told me was selenium-manager.exe had to be put in the following Windows folder:
```
"C:\Windows\System32\WindowsPowerShell\v1.0\selenium-manager\windows\selenium-manager.exe"
``` 
It didn't work if selenium-manager.exe was placed relative to webdriver.dll:
```
C:\Users\Steve\Documents\Backup\Task Scheduler Tasks\Library\SeleniumModule\selenium-manager\windows\selenium-manager.exe
``` 

I can see that it now downloads and accesses chromedriver.exe in my user .cache directory.

Thanks again.

**Comment by nvborisenko** [comment]
And this is strange, the location of selenium-manager.exe is predictable: https://github.com/SeleniumHQ/selenium/blob/5f3f7d096a75700e055effbe13cbc22117014e2c/dotnet/src/webdriver/SeleniumManager.cs#L46C16-L46C16

In any case it works for you, even if I have no explanation why :)

**Comment by titusfortner** [comment]
To be clear, Selenium Manager needs to work without having to load custom dll's. It should "just work." If it does not just work with the latest version of Selenium, please let us know.

**Comment by StevenTN615** [comment]
@titusfortner Not sure what you meant by "Selenium Manager needs to work without having to load custom dll's." Does that mean it should work without me having to put Newtonsoft.Json.dll in the same folder as Webdriver.dll? When I didn't have Newtonsoft.Json.dll I got the exception as [above](https://github.com/SeleniumHQ/selenium/issues/12564#issuecomment-1712403784).

**Comment by titusfortner** [comment]
Selenium comes with all the dlls, you don't need to download or install anything else. 

When you use Selenium 4.12.4 in your Selenium test, are you still getting an exception?

**Comment by StevenTN615** [comment]
Yes, I get the exception as [above](https://github.com/SeleniumHQ/selenium/issues/12564#issuecomment-1712403784).

Like I told @nvborisenko, I manually downloaded the 4.12.4 nupkg, extracted webdriver.dll to my working directory, and extracted selenium-manager.exe to "C:\Windows\System32\WindowsPowerShell\v1.0\selenium-manager\windows\selenium-manager.exe". When I run my PowerShell test script I get the "Could not load file or assembly 'Newtonsoft.Json, Version=13.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed' or one of its dependencies. The system cannot find the file specified.", exception. It only worked after I put Newtonsoft.Json.dll in the same directory as my Webdriver.dll.

**Comment by titusfortner** [comment]
Ah, I misunderstood, you are just debugging the problem locally. So for some reason Selenium is complaining about Newtonsoft.Json

**Comment by nuthanbanad** [comment]
> [nvborisenko](/nvborisenko)

I used latest version 4.12.4 and still got same exception. This is just a scheduled task on remote PC to run my app.exe which uses Selenium.

Below is the exception:

Unable to obtain chrome using Selenium Manager; For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors/driver_location
   at OpenQA.Selenium.DriverFinder.FullPath(DriverOptions options)
   at OpenQA.Selenium.Chromium.ChromiumDriver.GenerateDriverServiceCommandExecutor(DriverService service, DriverOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chromium.ChromiumDriver..ctor(ChromiumDriverService service, ChromiumOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeDriverService service, ChromeOptions options, TimeSpan commandTimeout)
   at OpenQA.Selenium.Chrome.ChromeDriver..ctor(ChromeOptions options)
   at DonwloadGroupDeviceListData.Form1.createSeleniumChromeBrowser()
   at DonwloadGroupDeviceListData.Form1.<Form1_Shown>d__25.MoveNext()

**Comment by StevenTN615** [comment]
@nuthanbanad What helped me was when I dumped the toString() method on my exception per @nvborisenko  suggestion. From that I was able to see 2 errors, selenium-manager.exe was in the wrong spot on my PC, and I needed Newtonsoft.Json.dll. Although @titusfortner says you shouldn't have to download any dependent dlls.

**Comment by nuthanbanad** [comment]
@StevenTN615 @nvborisenko 

Okay I have updated my code to show ex.toString(). I'll provide the exception details again if it occurs anytime soon as this issue is random for me.

All my .dlls are in same folder. Please see attached screenshot.
![Files](https://github.com/SeleniumHQ/selenium/assets/14852194/4e738441-9592-4afc-b271-97de1a373ec4)

**Comment by titusfortner** [comment]
@nvborisenko what is it that we need to do here?

**Comment by nvborisenko** [comment]
I think we can close it because of initially it was about 4.11. Many things have been improved, let's force people to post new issue instead of adding new comments here.
